xtra-libp2p/src/dialer.rs

```diff
@@ -96,18 +84,42 @@ impl Actor {
         self.listener_peer_id
             .expect("to always have peer id if successfully started")
     }
+
+    async fn is_connection_established(&self) -> Result<bool> {
+        Ok(self
+            .endpoint
+            .send(GetConnectionStats)
+            .await?
+            .connected_peers
+            .contains(&self.peer_id()))
+    }
+
+    async fn dial(&self) -> Result<()> {
+        anyhow::ensure!(
+            !self.is_connection_established().await?,
+            "Connection should not be active when dialing"
+        );
+
+        if let Err(e) = self.connect().await {
+            tracing::warn!("Failed to request connection from endpoint: {e:#}");
+        }
+
+        // Only check the connection again after it had enough time to be established
+        tokio::time::sleep(CONNECTION_TIMEOUT).await;
+
+        anyhow::ensure!(
```

Comment:
Should this really be an ensure? This changes the constraints. I would have seen it as a `warn` log if we could not establish the connection and I would keep a faster retry from the dialer (i.e. 5 secs)

Reply:
what's the changed constraint here? 

- the original behaviour: we wait, then we check, if we're not connected we stop the actor
- subscriber, v1: we spawn a task in the background that checks after the same timeout, and if we're not connected by then, we stop the actor.
- this PR: we wait, check, and if we're not connected we stop the actor (same thing really).

What changed a bit is how the actor is stopped, and that might have caused the confusion:
- original version: we dispatch an internal message to stop the actor
- subscribers, v1: we have a bunch of `if let Err(e)` that calls the `stop_with_error()` method
- this PR: I extracted an internal method, so I could use `if let Err(e)` once for the resulting anyhow error.

IMHO the second ensure is the crucial one, because this is the one that covers trying again. I don't see need for faster retry? (if you want, we can drop the alignment of the timeouts, so it always just tries every 5s, I can be convinced to that one :) )